### PR TITLE
Attacking exponential complexity in TypeMaps

### PR DESCRIPTION
**Pull request by retronym**
- Don't normalize existentials during the `contain`-s type map;
  `ExistentialType#normalize' calls contains internally and
  an exponential blowup ensues.
- Ensure that the type map used in variance validation never
  returns modified types in order to avoid needless cloning of
  symbols.

The enclosed test case still gets stuck in Uncurry, thanks to
the way that `TypeMap#mapOver(List[Symbol])` recurses through the
type first to check whether the type map would be an no-op or not.
If not, it repeats the type map with cloned symbols. Doing the work
twice at each level of recursion blows up the complexity.

Removing that "fast path" allows the enclosed test to compile
completely. As at this commit, it gets stuck in uncurry, which
dealiases `s.List` to `s.c.i.List` within the type.

Some more background on the troublesome part of `TypeMap`:

  http://lrytz.github.io/scala-aladdin-bugtracker/displayItem.do%3Fid=1210.html
  https://github.com/scala/scala/commit/f8b2b21050e7a2ca0f537ef70e3e0c8eead43abc

Review by @adriaanm

This is the uncontroversial piece of the change. I'm still trying out ideas to tackle the core complexity problem in `mapOver`.

### CODE REVIEW

**Comment by retronym** [comment]
Looks like dotty has the same idiom (check if the function is an identity, and then if not clone symbols and repeat) https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/core/Symbols.scala#L301-L335

/cc @odersky

**Comment by adriaanm** [comment]
Hello, 2007-me!

**Comment by retronym** [comment]
@adriaanm I'm clearing out the PR queue for 2.11.8, what to do you think about this one?

File: src/reflect/scala/reflect/internal/Variances.scala [hunk 1]
```diff
@@ -122,15 +122,18 @@ trait Variances {
...
+          case _ if isUncheckedVariance(tp)                    =>
+          case _ if resultTypeOnly(tp)                         => this(tp.resultType)
+          case TypeRef(_, sym, _) if sym.isAliasType           => this(tp.normalize)
+          case TypeRef(_, sym, _) if !sym.variance.isInvariant => checkVarianceOfSymbol(sym) ; mapOver(tp)
+          case RefinedType(_, _)                               => withinRefinement(mapOver(tp))
+          case ClassInfoType(parents, _, _)                    => parents foreach this
+          case mt @ MethodType(_, result)                      => flipped(mt.paramTypes foreach this) ; this(result)
+          case _                                               => mapOver(tp)
+        }
+        tp
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Variances.scala:139]: Maybe add a comment to explain why the result value is irrelevant? Or positively, why this should remain the identity?

**Comment by adriaanm** [comment]
Once the invariant of the variance map receives a memorable quote, this shall be very esthetically pleasing to me.

**Comment by retronym** [comment]
so commented.

**Comment by adriaanm** [comment]
LGTM -- splen-done

**Pull request merged**